### DISCUSSION TITLE
Is shallow (Instagram-style) routing on dynamic routes possible?

### DISCUSSION THREAD

**Discussion by adamwathan** [vercel/next.js#11625] [Help] [answered]
One thing I've never been able to get working is the Instagram-style "clicking a link opens a modal, hitting refresh shows a full page" routing behavior when working with dynamic routes.

Here's an example of a link from a demo project I was hacking on to try and get this working:

```js
<Link
  href={`/artists/${artist.id}/?showAlbum=${album.id}`}
  as={`/albums/${album.id}`}
  scroll={false}
  shallow
>
```

Here's a live demo:

https://headbangstagram-18x9bitrm.now.sh/artists/1

Notice that clicking an album cover goes to a new page instead of doing a shallow navigation (which should open a modal)?

Here's a direct link that includes the `showAlbum` query param, so you can see what should be happening:

https://headbangstagram-18x9bitrm.now.sh/artists/1?showAlbum=1

...and here's the source on GitHub for anyone interested:

https://github.com/adamwathan/headbangstagram/blob/master/pages/artists/%5Bid%5D.js#L58-L63

Is there any way to make this actually work with dynamic routes?

**Comment by Timer** [comment, marked as answer]
Hey!

A `<Link>`'s `href` must always be the route as it exists on the file system.

```js
// Incorrect (1):
<Link href="/artists/abc">

// Correct (2):
<Link href="/artists/[id]" as="/artists/abc" >

// Correct for Instagram-style route (3):
<Link href="/artists/[id]?id=abc" as="/different/page/on-refresh">
```

https://nextjs.org/docs/api-reference/next/link#dynamic-routes

You're free to mismatch the `href` and `as` for Instagram-style routes as long as you manually provide the necessary query information as shown in the above example (3). Otherwise, the route path is required to match for automatic detection (example 2 above).

I've sent you a PR fixing your demo: https://github.com/adamwathan/headbangstagram/pull/1
It should work as expected now!

**Reply by nandorojo** [reply]
When someone closes the modal, how would we reset the URL back to normal? Should I use a `shallow: true` route back to the same page? Thanks!

**Comment by adamwathan** [comment]
Super helpful thank you! Had a follow up question that I posted on the PR but probably makes more sense here since others might be interested, so pasting...

---

So the weird thing for me is that all of these work fine for regular links:

```html
<Link href="/projects/5" />
<Link href="/projects/[id]?id=5" as="/projects/5" />
<Link href="/projects/[id]" as="/projects/5" />
```

Why do they all work if not all of them are correct, and out of the last two options which one is the most correct?

**Reply by Timer** [reply]
The last two are functionally equivalent. Next.js automatically transforms the 3rd option into the 2nd — the 3rd is just the shorthand of the 2nd.

Regarding the 1st, this link must work for [zones support](https://nextjs.org/docs/advanced-features/multi-zones).

We need to show a warning when the `<Link>` doesn't resolve in development IMO, and we're doing that soon (n.b. you should see it's always a server-transition and not client-side render)!

**Comment by nimabrownlee** [comment]
Hey,
I wasn't able to get this to work when using `getStaticProps`: https://github.com/zeit/next.js/issues/11971
Any idea why? I used https://github.com/zeit/next.js/tree/canary/examples/with-route-as-modal.

**Reply by ptcampbell** [reply]
Having the same issue. Anyone else know?

**Reply by ptcampbell** [reply]
FWIW — I worked around it by using `swr` to get data instead of `getStaticProps`
